solana/indexer/payment_router.go

```diff
@@ -0,0 +1,161 @@
+package indexer
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"bridgerton.audius.co/config"
+	"bridgerton.audius.co/database"
+	"bridgerton.audius.co/solana/spl/programs/payment_router"
+	"github.com/gagliardetto/solana-go"
+	"github.com/jackc/pgx/v5"
+	"go.uber.org/zap"
+)
+
+func processPaymentRouterInstruction(
+	ctx context.Context,
+	db database.DBTX,
+	slot uint64,
+	tx *solana.Transaction,
+	instructionIndex int,
+	instruction solana.CompiledInstruction,
+	signature string,
+	blockTime time.Time,
+	config config.Config,
+	instLogger *zap.Logger,
+) error {
+
+	accounts, err := instruction.ResolveInstructionAccounts(&tx.Message)
+	if err != nil {
+		return fmt.Errorf("error resolving instruction accounts: %w", err)
+	}
+	inst, err := payment_router.DecodeInstruction(accounts, []byte(instruction.Data))
+	if err != nil {
+		return fmt.Errorf("error decoding payment_router instruction: %w", err)
+	}
+	switch inst.TypeID {
+	case payment_router.InstructionImplDef.TypeID(payment_router.Instruction_Route):
+		if routeInst, ok := inst.Impl.(*payment_router.Route); ok {
+			for i, account := range routeInst.GetDestinations() {
+				err = insertPayment(ctx, db, paymentRow{
+					signature:        signature,
+					instructionIndex: instructionIndex,
+					amount:           routeInst.Amounts[i],
+					slot:             slot,
+					routeIndex:       i,
+					toAccount:        account.PublicKey.String(),
+				})
+				if err != nil {
+					return fmt.Errorf("failed to insert payment at instruction: %w", err)
+				}
+			}
+
+			parsedPurchaseMemo, ok := findNextPurchaseMemo(tx, instructionIndex, instLogger)
+			if ok {
+				parsedLocationMemo := findNextLocationMemo(tx, instructionIndex, instLogger)
+				isValid, err := validatePurchase(ctx, config, db, routeInst, parsedPurchaseMemo, blockTime)
+				if err != nil {
+					instLogger.Error("invalid purchase", zap.Error(err))
+					// continue - insert the purchase as invalid for record keeping
+				}
+
+				err = insertPurchase(ctx, db, purchaseRow{
+					signature:          signature,
+					instructionIndex:   instructionIndex,
+					amount:             routeInst.TotalAmount,
+					slot:               slot,
+					fromAccount:        routeInst.GetSender().PublicKey.String(),
+					parsedPurchaseMemo: parsedPurchaseMemo,
+					parsedLocationMemo: parsedLocationMemo,
+					isValid:            isValid,
```

Comment:
I feel like this complicates things the way that `is_current` complicated things. Could we maybe just use a different table to track these so that we don't need a separate column we have to filter on for all but the edge cases?

Reply:
it definitely complicates things... do you feel the tradeoff is worth it though now that we chatted?

Reply:
If this is meant to enable the processing out of order case, then yes, probably. Otherwise it makes the indexing awkward because you have to check two tables for the purchase vs. just updating it from invalid->valid.